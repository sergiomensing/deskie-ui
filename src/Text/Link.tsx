import { AllHTMLAttributes, PropsWithChildren } from 'react';
import styled from 'styled-components';
import Text, { BaseTextProps } from './Text';

type LinkProps = AllHTMLAttributes<unknown> & BaseTextProps;

const StyledLink = styled(Text)`
  color: ${(props) => props.theme.colors.primary};

  &:hover {
    color: ${(props) => props.theme.colors.primaryHover};
  }
`;

function Link({ children, ...rest }: PropsWithChildren<LinkProps>) {
  return (
    <StyledLink as="a" {...rest}>
      {children}
    </StyledLink>
  );
}

export default Link;
