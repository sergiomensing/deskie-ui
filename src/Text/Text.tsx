import { HTMLAttributes, PropsWithChildren } from 'react';
import styled, { css, DefaultTheme, FlattenInterpolation, ThemeProps } from 'styled-components';

type TextAppearance = 'default' | 'success' | 'warning' | 'error' | 'faded' | 'secondary';

export type BaseTextProps = {
  p?: boolean;
  h1?: boolean;
  h2?: boolean;
  h3?: boolean;
  b?: boolean;
  small?: boolean;
  appearance?: TextAppearance;
  as?: keyof JSX.IntrinsicElements;
};

type TextProps = HTMLAttributes<unknown> & BaseTextProps;

type TextTags = Pick<TextProps, 'p' | 'h1' | 'h2' | 'h3' | 'small' | 'b'>;

type CssPartial = FlattenInterpolation<ThemeProps<DefaultTheme>>;

const StyledText = styled.span<{ appearance: TextAppearance; font: CssPartial }>`
  ${(props) => props.font};
  color: ${(props) =>
    props.appearance === 'default'
      ? props.theme.colors.foreground
      : props.appearance === 'secondary'
      ? props.theme.colors.accent8
      : props.appearance === 'faded'
      ? props.theme.colors.accent6
      : props.theme.colors[props.appearance]};
`;

function getElementTag({ p, h1, h2, h3 }: TextTags) {
  if (p) {
    return 'p';
  }

  if (h1) {
    return 'h1';
  }

  if (h2) {
    return 'h2';
  }

  if (h3) {
    return 'h3';
  }

  return 'span';
}

function getFont({ h1, h2, h3, small, b }: TextTags) {
  if (h1) {
    return css`
      font-size: 1.375rem;
      font-weight: 600;
      font-family: 'Poppins', sans-serif;
    `;
  }
  if (h2) {
    return css`
      font-size: 1.125rem;
      font-weight: 500;
      font-family: 'Poppins', sans-serif;
    `;
  }
  if (h3) {
    return css`
      font-size: 1rem;
      font-weight: 500;
    `;
  }
  if (small) {
    return css`
      font-size: 0.8125rem;
      font-weight: 400;
    `;
  }
  if (b) {
    return css`
      font-size: 0.875rem;
      font-weight: 600;
    `;
  }
  return css`
    font-size: 0.875rem;
    font-weight: 400;
    line-height: 1.75;
  `;
}

function Text({
  children,
  p,
  h1,
  h2,
  h3,
  small,
  b,
  as,
  appearance = 'default',
  ...rest
}: PropsWithChildren<TextProps>) {
  const tags = { p, h1, h2, h3, small, b };
  return (
    <StyledText as={as || getElementTag(tags)} font={getFont(tags)} appearance={appearance} {...rest}>
      {children}
    </StyledText>
  );
}

export default Text;
