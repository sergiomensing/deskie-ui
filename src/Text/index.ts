export { default as Text } from './Text';
export { default as Link } from './Link';
