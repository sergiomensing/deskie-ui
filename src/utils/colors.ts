export const colorShade = (col: string, amt: number) => {
  col = col.replace(/^#/, '');
  if (col.length === 3) col = col[0] + col[0] + col[1] + col[1] + col[2] + col[2];

  const [sr, sg, sb] = col.match(/.{2}/g) || ['ff', 'ff', 'ff'];
  const [ar, ag, ab] = [parseInt(sr, 16) + amt, parseInt(sg, 16) + amt, parseInt(sb, 16) + amt];

  const r = Math.max(Math.min(255, ar), 0).toString(16);
  const g = Math.max(Math.min(255, ag), 0).toString(16);
  const b = Math.max(Math.min(255, ab), 0).toString(16);

  const rr = (r.length < 2 ? '0' : '') + r;
  const gg = (g.length < 2 ? '0' : '') + g;
  const bb = (b.length < 2 ? '0' : '') + b;

  return `#${rr}${gg}${bb}`;
};

const defaultColors = [
  '#E64980',
  '#4C6EF5',
  '#228BE6',
  '#15AABF',
  '#12B886',
  '#40C057',
  '#82C91E',
  '#FAB005',
  '#FD7E14',
  '#FA5252',
];

export function getColorFromString(value: string, colors: undefined | string[] = defaultColors) {
  let hash = 0;
  if (value.length === 0) return colors[hash];
  for (let i = 0; i < value.length; i++) {
    hash = value.charCodeAt(i) + ((hash << 5) - hash);
    hash = hash & hash;
  }
  hash = ((hash % colors.length) + colors.length) % colors.length;
  return colors[hash];
}
