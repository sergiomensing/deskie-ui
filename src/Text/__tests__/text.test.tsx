import React from 'react';
import Text from '../Text';
import { render } from '../../../tests/utils';

describe('Text', () => {
  it('should render <span> element in the default', () => {
    const { container } = render(<Text>test-value</Text>);
    expect(container.querySelectorAll('span').length).not.toBe(0);
    expect(container).toMatchSnapshot();
  });

  it('should work with different appearances', () => {
    const { container } = render(
      <>
        <Text appearance="default">test-value-default</Text>
        <Text appearance="secondary">test-value-secondary</Text>
        <Text appearance="faded">test-value-faded</Text>
        <Text appearance="error">test-value-error</Text>
        <Text appearance="warning">test-value-warning</Text>
        <Text appearance="success">test-value-success</Text>
      </>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render the correct HTML tag', () => {
    const { container } = render(
      <>
        <Text h1>test-value-h1</Text>
        <Text h2>test-value-h2</Text>
        <Text h3>test-value-h3</Text>
        <Text p>test-value-paragraph</Text>
        <Text b>test-value-bold</Text>
        <Text small>test-value-small</Text>
      </>,
    );

    expect(container).toMatchSnapshot();
  });
});
