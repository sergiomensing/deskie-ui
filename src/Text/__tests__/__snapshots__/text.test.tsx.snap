// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text should render <span> element in the default 1`] = `
.c0 {
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 1.75;
  color: #11142D;
}

@media screen and (min-width:1024px) {

}

<div>
  <span
    class="c0"
  >
    test-value
  </span>
</div>
`;

exports[`Text should render the correct HTML tag 1`] = `
.c0 {
  font-size: 1.375rem;
  font-weight: 600;
  font-family: 'Poppins',sans-serif;
  color: #11142D;
}

.c1 {
  font-size: 1.125rem;
  font-weight: 500;
  font-family: 'Poppins',sans-serif;
  color: #11142D;
}

.c2 {
  font-size: 1rem;
  font-weight: 500;
  color: #11142D;
}

.c3 {
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 1.75;
  color: #11142D;
}

.c4 {
  font-size: 0.875rem;
  font-weight: 600;
  color: #11142D;
}

.c5 {
  font-size: 0.8125rem;
  font-weight: 400;
  color: #11142D;
}

@media screen and (min-width:1024px) {

}

<div>
  <h1
    class="c0"
  >
    test-value-h1
  </h1>
  <h2
    class="c1"
  >
    test-value-h2
  </h2>
  <h3
    class="c2"
  >
    test-value-h3
  </h3>
  <p
    class="c3"
  >
    test-value-paragraph
  </p>
  <span
    class="c4"
  >
    test-value-bold
  </span>
  <span
    class="c5"
  >
    test-value-small
  </span>
</div>
`;

exports[`Text should work with different appearances 1`] = `
.c0 {
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 1.75;
  color: #11142D;
}

.c1 {
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 1.75;
  color: #9395A1;
}

.c2 {
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 1.75;
  color: #D3D4DD;
}

.c3 {
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 1.75;
  color: #FF754C;
}

.c4 {
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 1.75;
  color: #FFCE73;
}

.c5 {
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 1.75;
  color: #7FBA7A;
}

@media screen and (min-width:1024px) {

}

<div>
  <span
    class="c0"
  >
    test-value-default
  </span>
  <span
    class="c1"
  >
    test-value-secondary
  </span>
  <span
    class="c2"
  >
    test-value-faded
  </span>
  <span
    class="c3"
  >
    test-value-error
  </span>
  <span
    class="c4"
  >
    test-value-warning
  </span>
  <span
    class="c5"
  >
    test-value-success
  </span>
</div>
`;
